content/blog/2020-12-18-db-compiler/index.md

```diff
@@ -0,0 +1,293 @@
+
+The key operators needed are Scan, Select, Join and Group By/Aggregate.
+
+Joins are implemented by creating a hash table of one of its inputs and then for each tuple of the other input, concatenating it with any matching tuple in the hash table. Similarly, grouping is done by maintaining a map of the group by values to a bucket that contains all values in the group. Then, each aggregation function is computed over each bucket.
+
+### Interpretation
+
+Most existing databases use interpretation to generate the output. This is generally implemented by having a produce function for each operator. When you call it, the operator will call a produce function from its child operators which returns the next tuple (or null if it's done), process that tuple (filter by some condition, execute a map function, join data, etc.), and then return the processed batch.
+
+For example, this would be what is executed for the interpreted version of `Filter(Scan(T), CompExpr(LT, ColRef(x), Literal(5)))`:
+```
+query {
+  output = []
+  while (t = child.produce()) {
+    output.add(t)
+  }
+  return output
+}
+
+ColRef.evaluate(t) {
+  return t[col]
+}
+
+CompExpr.evaluate(t) {
+   l = left.evaluate(t)
+   r = right.evaluate(t)
+   return op(l, r);
+}
+
+Filter.produce() {
+   while (t = child.produce()) {
+      if (expr.evaluate(t)) {
+        return t
+      }
+   }
+}
+
+Scan.produce() {
+  return next tuple of T
+}
+```
+
+While easy to implement from a system point of view, this leads to a lot of overhead as you have a function call boundary at each operator. This means values that are in registers need to be spilled and you have to invoke a virtual call which is expensive. 
+
+### Compilation
+Imagine the following equivalent code:
+```
+query {
+  output = []
+  for (t in T) {
+    if (t[x] > 5) {
+      output.add(t)
+    }
+  }
+  return output
+}
+```
+This should be much faster since there are no function calls (let alone virtual ones) and since tight loops are generated, values are kept in registered for as long as possible.
+
+## High Level Design
+
+### Producer/Consumer Interface
+
+To generate code for an operator tree, we introduce produce and consume functions for each operator. Produce  generates a loop over the tuples in an operator's result. Consume is called by an operator's child surrounded by a loop over its result tuples. Effectively, produce signifies to "pull" tuples from an operator while consume "push"es them. Expressions have a produce function that generate code for calculating that expression.
+
+To see this in action, let's consider the following psuedocode implementations for Scan and Filter:
+```
+Scan.produce() {
+  output << "for (tuple in T) {"
+    parent.consume()
+  output << "}"
+}
+
+Filter.produce() {
+  // To generate a loop over its result, generate a loop over its child's result.
+  child.produce();
+}
+
+Filter.consume() {
+  // This is called in a loop over the child's result
+  output << "if (";
+    cond.produce()
```

Comment:
Just checking, do conditions have `produce` functions too? Or should this be `evaluate`?

Reply:
Yep. Mentioned that above.